lego-powered-up/src/iodevice.rs

```diff
@@ -0,0 +1,249 @@
+        }
+    }
+    fn check(&self, mode: u8, datasettype: DatasetType) -> Result<()> {
+        if let Some(pm) = self.def.modes().get(&mode) {
+            let vf = pm.value_format;
+            // println!("Dataset for call: {:?}  Dataset for device: {:?} Device kind: {:?} ", &datasettype, &vf.dataset_type, &self.kind);
+            // dbg!(&self.modes);
+            if datasettype == vf.dataset_type {
+                Ok(())
+            } else {
+                Err(Error::NoneError(String::from("Incorrect dataset type")))
+            }
+        } else {
+            Err(Error::NoneError(String::from("Mode not found")))
+        }
+    }
+}
+
+impl RcDevice for IoDevice {
+    fn port(&self) -> u8 {
+        self.def.port()
+    }
+    fn tokens(&self) -> (&Peripheral, &Characteristic) {
+        (
+            (self.tokens.p.as_ref().unwrap()),
+            (self.tokens.c.as_ref().unwrap()),
+        )
+    }
+    fn get_rx_pvs(&self) -> Result<broadcast::Receiver<PortValueSingleFormat>> {
+        if let Some(sender) = &self.channels.singlevalue_sender {
+            Ok(sender.subscribe())
+        } else {
+            Err(Error::NoneError(String::from("Sender not found")))
+        }
+    }
+    fn get_rx_nwc(&self) -> Result<broadcast::Receiver<NetworkCommand>> {
+        if let Some(sender) = &self.channels.networkcmd_sender {
+            Ok(sender.subscribe())
+        } else {
+            Err(Error::NoneError(String::from("Sender not found")))
+        }
+    }
+    fn check(&self) -> Result<()> {
+        match self.def.kind() {
+            IoTypeId::RemoteButtons => Ok(()),
+            _ => Err(Error::HubError(String::from(
+                "Not a remote control device",
+            ))),
+        }
+    }
+}
+
+impl EncoderMotor for IoDevice {
+    fn port(&self) -> u8 {
+        self.def.port()
+    }
+    fn tokens(&self) -> (&Peripheral, &Characteristic) {
+        (
+            (self.tokens.p.as_ref().unwrap()),
+            (self.tokens.c.as_ref().unwrap()),
+        )
+    }
+    fn get_rx(&self) -> Result<broadcast::Receiver<PortValueSingleFormat>> {
+        if let Some(sender) = &self.channels.singlevalue_sender {
+            Ok(sender.subscribe())
+        } else {
+            Err(Error::NoneError(String::from("Sender not found")))
+        }
+    }
+    fn get_rx_combined(
+        &self,
+    ) -> Result<broadcast::Receiver<PortValueCombinedFormat>> {
+        if let Some(sender) = &self.channels.combinedvalue_sender {
+            Ok(sender.subscribe())
+        } else {
+            Err(Error::NoneError(String::from("Sender not found")))
+        }
+    }
+    fn check(&self) -> Result<()> {
+        match self.def.kind() {
+            IoTypeId::TechnicLargeLinearMotor
+            | IoTypeId::TechnicXLargeLinearMotor
+            | IoTypeId::InternalMotorTacho => Ok(()),
+            _ => Err(Error::HubError(String::from("Not an Encoder Motor"))),
+        }
```

Comment:
I've tested this with a TechnicMediumAngularMotorGrey from the 51515 Mindstorms set and it works
```suggestion
        match self.def.kind() {
            IoTypeId::TechnicLargeLinearMotor
            | IoTypeId::TechnicXLargeLinearMotor
            | IoTypeId::InternalMotorTacho
            | IoTypeId::TechnicMediumAngularMotorGrey => Ok(()),
            _ => Err(Error::HubError(String::from("Not an Encoder Motor"))),
        }
```